File: leetcode/balanced-binary-tree/balanced-binary-tree.cpp
/**************************************
* @author dujiong
* @date 2016.9.23
* @version V0.1
**************************************/

/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    bool isBalanced(TreeNode *root) {
        if(root==NULL){
            return true;
        }
        int leftDepth = depthOfBinaryTree(root->left);
        int rightDepth = depthOfBinaryTree(root->right);
        if(abs(leftDepth-rightDepth)<=1){
            if(isBalanced(root->left) && isBalanced(root->right)){
                return true;
            }
        }
        return false;
    }
private:
    int depthOfBinaryTree(TreeNode *root){
        if(root==NULL){
            return 0;
        }
        int ret = std::max(depthOfBinaryTree(root->left),depthOfBinaryTree(root->right));
        return ret+1;
    }
};

File: leetcode-golang/Balanced Binary Tree/Balanced Binary Tree.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
 func isBalanced(root *TreeNode) bool {
    if height(root) == -1 {
        return false
    }
    return true
}

func height(root *TreeNode) int {
    if root == nil {
        return 0
    }
    left := height(root.Left)
    right := height(root.Right)
    if left == -1 || right == -1 || math.Abs(float64(left)-float64(right)) > 1 {
        return -1
    }
    return int(math.Max(float64(left), float64(right)) + 1)
}
